ENH: `PhyloNode` and related files are now typed. Use `__slots__` on `PhyloNode`. Remove support for `TreeNode`.

rmcar17 opened the pull request:
Marked as draft as backwards compatibility needs to be added.

## Summary by Sourcery

Refactor tree implementation to use a fully typed PhyloNode class, remove legacy TreeNode support, and improve performance and correctness across tree operations and drawing utilities

New Features:
- Introduce __slots__ on PhyloNode for memory optimization
- Add comprehensive type annotations (Self, ParamSpec, precise return types) across PhyloNode and related modules
- Standardize on PhyloNode as the sole tree node class, dropping TreeNode

Enhancements:
- Refactor PhyloNode methods with typed parameters and overloads for list interfaces, enforcing parent/child link cleanup
- Streamline tree I/O by removing XML support and focusing on Newick/JSON serialization
- Clarify tree API parameter names (e.g. tip_name_1, tip_name_2) and refine drawing utilities with consistent type annotations
- Update configuration files (pyproject.toml, ruff.toml) for dependency formatting, linting rules, and staging

Tests:
- Update tests to use PhyloNode exclusively and accommodate new signatures and behaviors

## Summary by Sourcery

Migrate tree implementation to a fully typed PhyloNode class, remove legacy TreeNode support, and unify all tree operations and drawing utilities around PhyloNode. Introduce __slots__ and thorough type annotations, simplify serialization to Newick/JSON, revamp drawing geometry classes, update tests to match new interfaces, and refresh project configuration for dependency and linting formats.

New Features:
- Add __slots__ to PhyloNode for memory and performance optimization
- Introduce comprehensive type annotations (Self, ParamSpec, precise return types) in PhyloNode, drawing modules, and related code

Enhancements:
- Standardize on PhyloNode as the sole tree node class, dropping TreeNode and deprecating its support
- Refactor parent/child link management, copy and traversal methods in PhyloNode with typed interfaces
- Simplify tree I/O by removing XML write support and focusing on Newick and JSON formats
- Overhaul Dendrogram and TreeGeometry classes with consistent typing and improved coordinate calculations
- Consolidate tests to use PhyloNode exclusively and adapt to revised signatures and behaviors

Build:
- Update pyproject.toml dependency formatting, add plotly-stubs
- Refresh ruff.toml lint rules and per-file ignore directives

CI:
- Adjust ruff.toml for new lint settings

Tests:
- Revise tests throughout to replace TreeNode with PhyloNode and reflect updated APIs

rmcar17 commented on src/cogent3/core/tree.py:
```diff
@@ -400,30 +442,28 @@ def postorder(
         return curr
 
-    def rooted(self, edge_name: str) -> typing_extensions.Self:
+    def rooted(self, edge_name: str) -> Self:
```
If a tree is missing some branch lengths this might have weird behaviour.

rmcar17 commented on src/cogent3/core/tree.py:
```diff
@@ -782,10 +798,12 @@ def prune(self, keep_root: bool = False) -> None:
                     node_val = node.params.get(key)
                     if is_number(value) or is_number(node_val):
-                        value = (value or 0.0) + (node_val or 0.0)
+                        value = (value or 0.0) + (node_val or 0.0)  # noqa: PLW2901
```
Also this only runs when branch lengths exist so unclear for the original intention is.

rmcar17 commented on src/cogent3/core/tree.py:
```diff
@@ -1508,14 +1512,19 @@ def tip_to_tip_distances(
                 default_length=default_length,
             )
-        default_length = 0 if hasattr(self, "length") else 1
+        default_length = (
```
This doesn't respect the original default_length - unclear why something like this was here originally

rmcar17 commented on src/cogent3/core/tree.py:
```diff
@@ -507,13 +543,11 @@ def iter_tips(self, include_self=False):
                 yield curr
 
-    def tips(self, include_self=False):
+    def tips(self, include_self: bool = False) -> list[Self]:
```
Should this default to True?

GavinHuttley replied:
I think it's intended to return tip descendants from self. If self is a tip, I'd expect that result to be an empty list. I think we need to discuss.

rmcar17 replied:
I see both schools of thought, I suppose there is now a divergence between get_tip_names and tips

GavinHuttley commented on src/cogent3/core/tree.py:
```diff
@@ -51,33 +57,25 @@
-
+    from typing_extensions import Self
 
-def distance_from_r(m1, m2):
```
Yes, always deprecate public methods. And add the   `# pragma: no cover` tag.

GavinHuttley commented on src/cogent3/core/tree.py:
```diff
@@ -189,129 +204,159 @@ def source(self, value: str | None) -> None:
     __deepcopy__ = deepcopy = copy
 
-    def copy_topology(self, constructor=None):
+    def copy_topology(self, constructor: type[T] | None = None) -> T | Self:
```
Agreed. And with the `deprecate_args` decorator, in a case like this where it's not being used, you can just delete it from the signature.

rmcar17 replied:
Decided to discontinue and instead rely on .copy()

GavinHuttley commented on src/cogent3/core/tree.py:
```diff
@@ -189,129 +204,159 @@ def source(self, value: str | None) -> None:
+        return result
+
     # support for copy module
-    def copy(self, memo=None, _nil=None, constructor: None = None):
```
deprecate as before

GavinHuttley commented on src/cogent3/core/tree.py:
```diff
@@ -1286,18 +1277,22 @@ def get_node_names(
         return [node.name for node in nodes]
 
-    def get_tip_names(self, include_self: bool = False) -> list[str]:
+    def get_tip_names(self, include_self: bool = True) -> list[str]:
```
Agreed.

GavinHuttley commented on src/cogent3/draw/dendrogram.py:
```diff
@@ -818,22 +864,23 @@ def _build_fig(self, **kwargs) -> None:
-            self.layout.pop("shapes", None)
+            self.layout.pop("shapes", None)  # type: ignore[reportUnknownMemberType]
 
+        # TODO: look into TypedDict as all the type: ignores are getting a bit silly
```
The problem is UnionDict is really being used as a "propertied" dict (keys are asccessible as properties) so it mimics the behaviour of plotly's Layout class. The reason we did that was to avoid a massive overhead of importing Plotly until truly needing it for display.
I agree this decision needs to be revisited, but it's not a priority.

rmcar17 commented on src/cogent3/core/tree.py:
```diff
@@ -29,20 +29,26 @@
 
-import typing_extensions
-from numpy import ceil, log, zeros
+import numpy as np
```
fwiw np isn't used in the codebase anywhere for the number of processes and treating numpy as np is pretty standard across the python ecosystem, but have made the change to numpy

rmcar17 commented on tests/test_parse/test_tree.py:
```diff
@@ -130,32 +130,32 @@ class DndParserTests(TestCase):
-        exp.children[0].append(PhyloNode())
-        exp.children[1].append(PhyloNode())
-        exp.children[1].append(PhyloNode())
+        exp = PhyloNode("")
```
Could do, however I feel like nodes not having names should be the exception rather than the rule and default args might encourage users to not specify names. In the tree contains any duplicate names then it breaks functions like `to_rich_dict`, `get_nodes_dict`, perhaps `subsets`, `child_parent_map`, and worse if duplicate tip names are allowed. An alternative approach might be to have a default of None and move the `TreeBuilder._unique_name` behaviour to `PhyloNode` and activate this on None.

rmcar17 commented:
Can't reply to the bool array suggestion but done that now

rmcar17 commented:
Just updated the PR addressing a large batch of the feedback, and some more wide-reaching changes making length/support attributes. Added a new key called `length_and_support` to `PhyloNode.to_rich_dict` in case users want to support other meanings of length/support in the params dict (which would otherwise cause a conflict here).

rmcar17 commented:
@GavinHuttley now passes all tests

GavinHuttley commented on src/cogent3/core/tree.py:
```diff
@@ -2263,13 +2369,29 @@ def load_tree(
+        "dict[str, dict[str, float | None]] | None", data.get("length_and_support")
+    )
+
+    if length_and_support is None:
```
As the rich dict method is used for json serialisation we need to provide an update mechanism, which means this deserialise function should handle the two versions. In the old version, there is only the `"edge_attributes"` key. We can check the "version" string in `data`, and/or just do the following
```python
if "length_and_support" not in data:
    length_and_support = {}
    for edge_name, edge_params in edge_attr.items():
          length_and_support[edge_name] = {"length": edge_params.pop("length", None), "support": edge_params.pop("support", None)}
```
Then a test case can be based on
```python
{'newick': '(Galago,(Gorilla,Rhesus)edge.0,HowlerMon)root',
 'edge_attributes': {'Galago': {'length': 0.1417910624295473},
  'Gorilla': {'length': 0.0173247791826725},
  'Rhesus': {'length': 0.020457707345485687},
  'edge.0': {'length': 0.0073129236698150635},
  'HowlerMon': {'length': 0.04134744219481945},
  'root': {'length': None}},
 'type': 'cogent3.core.tree.PhyloNode',
 'version': '2025.7.10a6'}
```

GavinHuttley commented on src/cogent3/core/tree.py:
```diff
@@ -2263,13 +2369,29 @@ def load_tree(
+    )
+
+    if length_and_support is None:
+        warnings.warn(
```
then delete this warning

GavinHuttley reviewed (changes requested):
Yet another monumental effort!

The only remaining issue is that the deserialise function should handle both the old and the new types. I've added some code and some data for testing that you can use to address this as a comment.

rmcar17 commented:
@GavinHuttley actually the deserialiser already handles the new and old json formats. It does this by checking for the existence of the `"length_and_support"` key and if it's not there then it extracts the length/support from the params dict. I included the warning there to suggest to people to migrate to the new format.

GavinHuttley commented:
@rmcar17 and of course you're right! I should have been a bit more careful in my reading. It doesn't look like the two versions are tested though?

rmcar17 commented:
Will add test

rmcar17 commented:
@GavinHuttley added a test checking whether loading from the old format works as expected!

GavinHuttley reviewed (approved):
Awesome job @rmcar17 !
